File: ts-portal-db/app/schemas/customer.py
"""
고객사 관련 Pydantic 스키마
"""

from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, EmailStr, Field


class CustomerBase(BaseModel):
    """고객사 기본 스키마"""
    company_name: str = Field(..., min_length=1, max_length=100, description="고객사명")
    contact_person: Optional[str] = Field(None, max_length=50, description="담당자명")
    contact_email: Optional[str] = Field(None, description="담당자 이메일")
    contact_phone: Optional[str] = Field(None, max_length=20, description="담당자 전화번호")
    contract_type: Optional[str] = Field(None, max_length=50, description="계약 유형")
    contract_start: Optional[date] = Field(None, description="계약 시작일")
    contract_end: Optional[date] = Field(None, description="계약 종료일")
    notes: Optional[str] = Field(None, description="특이사항 및 메모")


class CustomerCreate(CustomerBase):
    """고객사 생성 스키마"""
    status: str = Field(default="Active", description="상태")


class CustomerUpdate(BaseModel):
    """고객사 수정 스키마"""
    company_name: Optional[str] = Field(None, min_length=1, max_length=100, description="고객사명")
    contact_person: Optional[str] = Field(None, max_length=50, description="담당자명")
    contact_email: Optional[str] = Field(None, description="담당자 이메일")
    contact_phone: Optional[str] = Field(None, max_length=20, description="담당자 전화번호")
    contract_type: Optional[str] = Field(None, max_length=50, description="계약 유형")
    contract_start: Optional[date] = Field(None, description="계약 시작일")
    contract_end: Optional[date] = Field(None, description="계약 종료일")
    status: Optional[str] = Field(None, description="상태")
    notes: Optional[str] = Field(None, description="특이사항 및 메모")


class CustomerResponse(CustomerBase):
    """고객사 응답 스키마"""
    id: int
    status: str
    is_active: bool = Field(description="계약 활성 상태")
    contract_days_remaining: Optional[int] = Field(description="계약 종료까지 남은 일수")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

    @classmethod
    def from_orm_with_calculated(cls, customer):
        """ORM 객체에서 계산된 필드 포함하여 변환"""
        data = {
            "id": customer.id,
            "company_name": customer.company_name,
            "contact_person": customer.contact_person,
            "contact_email": customer.contact_email,
            "contact_phone": customer.contact_phone,
            "contract_type": customer.contract_type,
            "contract_start": customer.contract_start,
            "contract_end": customer.contract_end,
            "status": customer.status,
            "notes": customer.notes,
            "is_active": customer.is_active,
            "contract_days_remaining": customer.contract_days_remaining,
            "created_at": customer.created_at,
            "updated_at": customer.updated_at
        }
        return cls(**data)


class CustomerListResponse(BaseModel):
    """고객사 목록 응답 스키마"""
    total: int
    customers: List[CustomerResponse] 

File: frontend-new/src/types/customer.ts
/**
 * Customer 관련 타입 정의
 */

// 고객사 상태 enum (백엔드와 동일)
export type CustomerStatus = 'Active' | 'Pending' | 'Expired' | 'Suspended'

// 담당자 역할 (백엔드와 동일)
export type AssignmentRole = 'Primary' | 'Secondary' | 'Backup'

// 고객사 기본 정보 (백엔드 스키마와 일치)
export interface Customer {
  id: number
  company_name: string
  contact_person?: string
  contact_email?: string
  contact_phone?: string
  contract_type?: string
  contract_start?: string
  contract_end?: string
  status: string
  notes?: string
  is_active: boolean
  contract_days_remaining?: number
  created_at: string
  updated_at: string
  assignments?: Assignment[]
}

// 고객사 생성 요청 (백엔드 스키마와 일치)
export interface CustomerCreate {
  company_name: string
  contact_person?: string
  contact_email?: string
  contact_phone?: string
  contract_type?: string
  contract_start?: string
  contract_end?: string
  status?: CustomerStatus
  notes?: string
}

// 고객사 수정 요청 (백엔드 스키마와 일치)
export interface CustomerUpdate {
  company_name?: string
  contact_person?: string
  contact_email?: string
  contact_phone?: string
  contract_type?: string
  contract_start?: string
  contract_end?: string
  status?: CustomerStatus
  notes?: string
}

// 담당자 배정 정보 (백엔드 스키마와 일치)
export interface Assignment {
  id: number
  member_id: number
  customer_id: number
  role: AssignmentRole
  assigned_date: string
  end_date?: string
  is_primary: boolean
  is_active: boolean
  responsibilities?: string
  duration_days: number
  created_at: string
  updated_at: string
  member?: MemberInfo
  customer?: Customer
}

// 담당자 배정 생성 요청
export interface AssignmentCreate {
  member_id: number
  customer_id: number
  role?: AssignmentRole
  is_primary?: boolean
  responsibilities?: string
  end_date?: string
}

// 담당자 배정 수정 요청
export interface AssignmentUpdate {
  role?: AssignmentRole
  is_primary?: boolean
  responsibilities?: string
  end_date?: string
  is_active?: boolean
}

// API 응답 타입들 (백엔드 스키마와 일치)
export interface CustomerResponse extends Customer {}

export interface CustomerListResponse {
  total: number
  customers: Customer[]
}

export interface AssignmentResponse extends Assignment {}

export interface AssignmentListResponse {
  assignments: Assignment[]
  total: number
  skip: number
  limit: number
}

// 팀원 정보 (다른 서비스에서 가져온 정보)
export interface MemberInfo {
  id: number
  name: string
  email: string
  position?: string
  team?: string
}

// 검색 및 필터 파라미터
export interface CustomerSearchParams {
  q?: string
  status?: CustomerStatus
  active_only?: boolean
  contract_type?: string
  skip?: number
  limit?: number
}

export interface AssignmentSearchParams {
  member_id?: number
  customer_id?: number
  active_only?: boolean
  skip?: number
  limit?: number
}

// 통계 정보 (백엔드 스키마와 일치)
export interface CustomerStats {
  total_customers: number
  active_customers: number
  inactive_customers: number
  expired_customers: number
  expiring_soon: number
  active_rate: number
}

export interface AssignmentStats {
  total_assignments: number
  active_assignments: number
  primary_assignments: number
  secondary_assignments: number
  backup_assignments: number
  assignments_by_member: Record<string, number>
} 